### PR TITLE
`TopAppBar` refactorings

### LINKED ISSUE #212: Update scrollable TopAppBar to Material3-like API
- separate state and behavior builders (https://github.com/androidx/androidx/commit/e260141f2a679c33bfbe576d3a73fd4338d54d1a)
- scroll (drag) on TopAppBar also scrolls (https://github.com/androidx/androidx/commit/b925baec97c19cb542f90318091ab2d66bc448c5)
- scroll rework (https://github.com/androidx/androidx/commit/177e38797cc4f9749704170a0c04f838117de3b2)

### PR DESCRIPTION
**Pull request by hrach**
Closes #212. Backported all Material 3 changes.

### CODE REVIEW

File: ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBarScrollBehavior.kt [hunk 1]
```diff
@@ -293,102 +293,238 @@ private class ExitUntilCollapsedScrollBehavior(
...
+     * This value is updated by a [TopAppBarScrollBehavior] whenever a nested scroll connection
+     * consumes scroll events. A common implementation would update the value to be the sum of all
+     * [NestedScrollConnection.onPostScroll] `consumed.y` values.
+     */
+    public var contentOffset: Float by mutableStateOf(initialContentOffset)
+
+    /**
+     * A value that represents the collapsed height percentage of the app bar.
+     *
+     * A `0.0` represents a fully expanded bar, and `1.0` represents a fully collapsed bar (computed
```
Comment (Drekorian) [inline ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBarScrollBehavior.kt:484]: Technically, `0.0` and `1.0` are `Double` literals, perhaps replace with `0.0f` and `1.0f`, respectively, for clarity.

File: ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBarScrollBehavior.kt [hunk 2]
```diff
@@ -293,102 +293,238 @@ private class ExitUntilCollapsedScrollBehavior(
...
+            heightOffset / heightOffsetLimit
+        } else {
+            0f
+        }
+
+    /**
+     * A value that represents the percentage of the app bar area that is overlapping with the
+     * content scrolled behind it.
+     *
+     * A `0.0` indicates that the app bar does not overlap any content, while `1.0` indicates that
```
Comment (Drekorian) [inline ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBarScrollBehavior.kt:498]: Dtto.

**Review by Drekorian — APPROVED** [review]

File: ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBar.kt [hunk 3]
```diff
@@ -121,8 +124,28 @@ private fun SingleRowTopAppBar(
     extraContent: @Composable () -> Unit,
     scrollBehavior: TopAppBarScrollBehavior?,
 ) {
+    // Set up support for resizing the top app bar when vertically dragging the bar itself.
+    val appBarDragModifier = if (scrollBehavior != null && !scrollBehavior.isPinned) {
```
[the remark this answers is not in the document]
Comment (hrach) [reply ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBar.kt:128]: Fixed, this is more "our" code than other stuff.

File: ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBarScrollBehavior.kt [hunk 4]
```diff
@@ -25,139 +25,139 @@ import kotlin.math.abs
...
+                canScroll = canScroll,
+            )
 
+        /**
+         * Returns a [TopAppBarScrollBehavior] that adjusts its properties to affect the colors and
+         * height of the top app bar.
+         *
+         * A top app bar that is set up with this [TopAppBarScrollBehavior] will immediately collapse
+         * when the nested content is pulled up, and will expand back the collapsed area when the
+         * content is  pulled all the way down.
```
[the remark this answers is not in the document]
Comment (hrach) [reply ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBarScrollBehavior.kt:113]: copy-pasta. keeping

File: ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBarScrollBehavior.kt [hunk 5]
```diff
@@ -185,35 +185,30 @@ private class PinnedScrollBehavior(val canScroll: () -> Boolean = { true }) :
  * A top app bar that is set up with this [TopAppBarScrollBehavior] will immediately collapse when
  * the nested content is pulled up, and will immediately appear when the content is pulled down.
  *
+ * @param state a [TopAppBarState]
+ * @param snapAnimationSpec an optional [AnimationSpec] that defines how the top app bar snaps to
+ * either fully collapsed or fully extended state when a fling or a drag scrolled it into an
```
[the remark this answers is not in the document]
Comment (hrach) [reply ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBarScrollBehavior.kt:190]: copy-pasta. keeping

File: ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBarScrollBehavior.kt (hunk 1 above)
Comment (hrach) [reply ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBarScrollBehavior.kt:484]: copy-pasta. keeping

File: ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBarScrollBehavior.kt (hunk 2 above)
Comment (hrach) [reply ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBarScrollBehavior.kt:498]: copy-pasta. keeping

File: ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBarScrollBehavior.kt [hunk 6]
```diff
@@ -252,34 +254,32 @@ private class EnterAlwaysScrollBehavior(val canScroll: () -> Boolean = { true })
  * the nested content is pulled up, and will expand back the collapsed area when the content is
  * pulled all the way down.
  *
- * @param decayAnimationSpec a [DecayAnimationSpec] that will be used by the top app bar motion
- * when the user flings the content. Preferably, this should match the animation spec used by the
- * scrollable content. See also [androidx.compose.animation.rememberSplineBasedDecay] for a
- * default [DecayAnimationSpec] that can be used with this behavior.
+ * @param state a [TopAppBarState]
+ * @param snapAnimationSpec an optional [AnimationSpec] that defines how the top app bar snaps to
+ * either fully collapsed or fully extended state when a fling or a drag scrolled it into an
```
[the remark this answers is not in the document]
Comment (hrach) [reply ui/src/main/java/kiwi/orbit/compose/ui/controls/TopAppBarScrollBehavior.kt:259]: copy-pasta. keeping

**Pull request merged**